src/robot/running/arguments/typeconverters.py

```diff
@@ -389,61 +389,165 @@ class ListConverter(TypeConverter):
     type_name = 'list'
     abc = Sequence
     value_types = (str, Sequence)
+    aliases = ('list',)
 
-    def no_conversion_needed(self, value):
-        if isinstance(value, str):
+    def __init__(self, type_, custom_converters=None, languages=None):
+        super().__init__(type_, languages=languages)
+        self.nested_types = getattr(type_, '__args__', ())
+
+    @classmethod
+    def handles(cls, type_):
+        base_type = getattr(type_, '__origin__', type_)
+        handled = (cls.type, cls.abc) if cls.abc else cls.type
+        if base_type is tuple:
             return False
-        return super().no_conversion_needed(value)
+        return isinstance(base_type, type) and issubclass(base_type, handled)
+
+    def no_conversion_needed(self, value):
+        return not self.nested_types \
+               and isinstance(value, self.used_type) \
+               and not isinstance(value, str)
 
     def _non_string_convert(self, value, explicit_type=True):
-        return list(value)
+        if isinstance(value, str):
+            converted_list = self._literal_eval(value, list)
+        elif is_list_like(value):
+            converted_list = list(value)
+        else:
+            raise ValueError
+        if self.nested_types:
+            if TypeConverter.converter_for(self.nested_types[0],
+                                           languages=self.languages):
+                for i, elem in enumerate(converted_list) :
+                    converter = TypeConverter.converter_for(self.nested_types[0],
+                                                            languages=self.languages)
+                    if converter:
+                        elem_info = f"List[{type_name(self.nested_types[0])}]"
+                        converted_list[i] = converter.convert(elem_info, elem,
```

Comment:
This code seems to be responsible on that 
> Argument 'List[integer]' got value '3.14' (float) that cannot be converted to integer

error message that is a bit confusing because the actual argument is the whole list. I think the error could be improved by catching the exception raised by `converter.convert` and then re-raising it with more info. Good error messages are very important, but I believe it's better to get other code cleaned up before looking at this too much.

Reply:
I agree that good error messages are very important. Something was mentioned on this elsewhere as well. Let's continue the discussion there.